File: src/CircleAI.Markets/InMemoryMarkets.cs
// InMemoryMarkets.cs
//
// (3.3.0) Real in-memory market-data feed + instrument catalog + order
// router. The feed supports subscribe/broadcast quote pushes; the
// order router accepts and rejects based on simple rules (positive
// quantity, known instrument, valid limit price for limit orders).

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Markets;

public sealed class InMemoryInstrumentCatalog : IInstrumentCatalog
{
    private readonly ConcurrentDictionary<string, Instrument> _items = new(StringComparer.OrdinalIgnoreCase);

    public string BackendId => "in-memory";

    public void Add(Instrument item) => _items[item.Symbol] = item ?? throw new ArgumentNullException(nameof(item));

    public ValueTask<Instrument?> GetAsync(string symbol, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(symbol)) throw new ArgumentException("symbol required", nameof(symbol));
        _items.TryGetValue(symbol, out var i);
        return ValueTask.FromResult(i);
    }

    public ValueTask<IReadOnlyList<Instrument>> SearchAsync(string query, int topK = 20, CancellationToken ct = default)
    {
        if (query is null) throw new ArgumentNullException(nameof(query));
        if (topK <= 0) throw new ArgumentOutOfRangeException(nameof(topK));
        var hits = _items.Values
            .Where(i => i.Symbol.Contains(query, StringComparison.OrdinalIgnoreCase))
            .OrderBy(i => i.Symbol)
            .Take(topK)
            .ToArray();
        return ValueTask.FromResult<IReadOnlyList<Instrument>>(hits);
    }
}

public sealed class InMemoryMarketDataFeed : IMarketDataFeed
{
    private readonly ConcurrentDictionary<string, Quote> _quotes = new(StringComparer.OrdinalIgnoreCase);
    private readonly ConcurrentDictionary<string, List<Func<Quote, ValueTask>>> _subs = new(StringComparer.OrdinalIgnoreCase);
    private readonly object _gate = new();

    public string BackendId => "in-memory";

    public void Publish(Quote q)
    {
        ArgumentNullException.ThrowIfNull(q);
        _quotes[q.Symbol] = q;
        if (_subs.TryGetValue(q.Symbol, out var list))
        {
            Func<Quote, ValueTask>[] snap;
            lock (_gate) snap = list.ToArray();
            foreach (var s in snap)
            {
                try { _ = s(q); }
                catch (Exception ex) { System.Diagnostics.Debug.WriteLine($"[CircleAI.Markets] quote subscriber threw: {ex.Message}"); }
            }
        }
    }

    public ValueTask<Quote?> GetQuoteAsync(string symbol, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(symbol)) throw new ArgumentException("symbol required", nameof(symbol));
        _quotes.TryGetValue(symbol, out var q);
        return ValueTask.FromResult(q);
    }

    public IDisposable SubscribeQuotes(string symbol, Func<Quote, ValueTask> handler)
    {
        if (string.IsNullOrWhiteSpace(symbol)) throw new ArgumentException("symbol required", nameof(symbol));
        ArgumentNullException.ThrowIfNull(handler);
        var list = _subs.GetOrAdd(symbol, _ => new List<Func<Quote, ValueTask>>());
        lock (_gate) list.Add(handler);
        return new Subscription(this, symbol, handler);
    }

    private sealed class Subscription : IDisposable
    {
        private readonly InMemoryMarketDataFeed _owner;
        private readonly string _sym;
        private readonly Func<Quote, ValueTask> _h;
        public Subscription(InMemoryMarketDataFeed o, string s, Func<Quote, ValueTask> h) { _owner = o; _sym = s; _h = h; }
        public void Dispose()
        {
            if (_owner._subs.TryGetValue(_sym, out var list))
            {
                lock (_owner._gate) list.Remove(_h);
            }
        }
    }
}

public sealed class InMemoryOrderRouter : IOrderRouter
{
    private readonly IInstrumentCatalog _catalog;
    private long _seq;

    public InMemoryOrderRouter(IInstrumentCatalog catalog)
        => _catalog = catalog ?? throw new ArgumentNullException(nameof(catalog));

    public string BackendId => "in-memory";

    public async ValueTask<OrderResult> SubmitAsync(OrderRequest req, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(req);
        if (req.Quantity <= 0)
            return new OrderResult(NextId(), false, "Quantity must be positive");
        if (req.Type == OrderType.Limit && (req.LimitPrice is null || req.LimitPrice <= 0))
            return new OrderResult(NextId(), false, "Limit order requires positive LimitPrice");

        var inst = await _catalog.GetAsync(req.Symbol, ct).ConfigureAwait(false);
        if (inst is null)
            return new OrderResult(NextId(), false, "Unknown symbol");

        return new OrderResult(NextId(), true, null);
    }

    private string NextId() => $"ord-{Interlocked.Increment(ref _seq)}";
}

File: python/src/circle_ai/markets/in_memory_markets.py
# in_memory_markets.py
#
# Port of CircleAI.Markets InMemoryMarkets.cs (C# — the EXACT spec).
#
# (3.3.0) Real in-memory market-data feed + instrument catalog + order router.
# The feed supports subscribe/broadcast quote pushes; the order router accepts
# and rejects based on simple rules (positive quantity, known instrument, valid
# limit price for limit orders).
#
# C# ConcurrentDictionary(StringComparer.OrdinalIgnoreCase) maps to a plain dict
# keyed by the case-folded symbol (so lookups are case-insensitive) that stores
# the original record. C# ValueTask methods map to async def. A monotonically
# increasing order id "ord-{n}" is minted from a lock-guarded counter.
#
# Publish is a *sync* fan-out: it snapshots the subscriber list under the gate
# and invokes each async handler OUTSIDE the gate (so a handler that re-enters
# subscribe/dispose cannot self-deadlock), scheduling the returned coroutine
# fire-and-forget and swallowing any exception — mirroring the C# `_ = s(q)` in a
# try/catch. This preserves the "a throwing subscriber never breaks Publish"
# contract of the C# feed.

from __future__ import annotations

import asyncio
import threading
from typing import Dict, List, Optional

from .contracts import (
    IDisposable,
    IInstrumentCatalog,
    IMarketDataFeed,
    IOrderRouter,
    Instrument,
    OrderRequest,
    OrderResult,
    OrderType,
    Quote,
    QuoteHandler,
)


class InMemoryInstrumentCatalog(IInstrumentCatalog):
    """Thread-safe in-memory :class:`IInstrumentCatalog` (case-insensitive
    symbol keys).
    """

    def __init__(self) -> None:
        self._items: Dict[str, Instrument] = {}
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    def add(self, item: Instrument) -> None:
        if item is None:
            raise ValueError("item must not be None")
        with self._lock:
            self._items[item.symbol.casefold()] = item

    async def get_async(
        self, symbol: str, ct: Optional[object] = None
    ) -> Optional[Instrument]:
        if symbol is None or not symbol.strip():
            raise ValueError("symbol required")
        with self._lock:
            return self._items.get(symbol.casefold())

    async def search_async(
        self, query: str, top_k: int = 20, ct: Optional[object] = None
    ) -> List[Instrument]:
        if query is None:
            raise ValueError("query must not be None")
        if top_k <= 0:
            raise ValueError("top_k")
        q = query.casefold()
        with self._lock:
            hits = [i for i in self._items.values() if q in i.symbol.casefold()]
        # C# OrderBy(i => i.Symbol) — ordinal (case-sensitive) ascending.
        hits.sort(key=lambda i: i.symbol)
        return hits[:top_k]


class InMemoryMarketDataFeed(IMarketDataFeed):
    """Thread-safe in-memory :class:`IMarketDataFeed` with subscribe/broadcast."""

    def __init__(self) -> None:
        self._quotes: Dict[str, Quote] = {}
        self._subs: Dict[str, List[QuoteHandler]] = {}
        self._gate = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    def publish(self, q: Quote) -> None:
        if q is None:
            raise ValueError("quote must not be None")
        key = q.symbol.casefold()
        with self._gate:
            self._quotes[key] = q
            snap = list(self._subs.get(key, ()))
        for handler in snap:
            _fire(handler, q)

    async def get_quote_async(
        self, symbol: str, ct: Optional[object] = None
    ) -> Optional[Quote]:
        if symbol is None or not symbol.strip():
            raise ValueError("symbol required")
        with self._gate:
            return self._quotes.get(symbol.casefold())

    def subscribe_quotes(self, symbol: str, handler: QuoteHandler) -> IDisposable:
        if symbol is None or not symbol.strip():
            raise ValueError("symbol required")
        if handler is None:
            raise ValueError("handler must not be None")
        key = symbol.casefold()
        with self._gate:
            self._subs.setdefault(key, []).append(handler)
        return _Subscription(self, key, handler)

    def _unsubscribe(self, key: str, handler: QuoteHandler) -> None:
        with self._gate:
            lst = self._subs.get(key)
            if lst is not None:
                try:
                    lst.remove(handler)
                except ValueError:
                    pass


class _Subscription(IDisposable):
    def __init__(
        self, owner: InMemoryMarketDataFeed, key: str, handler: QuoteHandler
    ) -> None:
        self._owner = owner
        self._key = key
        self._handler = handler
        self._disposed = False
        self._lock = threading.Lock()

    def dispose(self) -> None:
        with self._lock:
            if self._disposed:
                return
            self._disposed = True
        self._owner._unsubscribe(self._key, self._handler)


class InMemoryOrderRouter(IOrderRouter):
    """In-memory :class:`IOrderRouter` — accepts/rejects on simple rules."""

    def __init__(self, catalog: IInstrumentCatalog) -> None:
        if catalog is None:
            raise ValueError("catalog must not be None")
        self._catalog = catalog
        self._seq = 0
        self._lock = threading.Lock()

    @property
    def backend_id(self) -> str:
        return "in-memory"

    async def submit_async(
        self, req: OrderRequest, ct: Optional[object] = None
    ) -> OrderResult:
        if req is None:
            raise ValueError("req must not be None")
        if req.quantity <= 0:
            return OrderResult(self._next_id(), False, "Quantity must be positive")
        if req.type == OrderType.Limit and (
            req.limit_price is None or req.limit_price <= 0
        ):
            return OrderResult(
                self._next_id(), False, "Limit order requires positive LimitPrice"
            )
        inst = await self._catalog.get_async(req.symbol, ct)
        if inst is None:
            return OrderResult(self._next_id(), False, "Unknown symbol")
        return OrderResult(self._next_id(), True, None)

    def _next_id(self) -> str:
        with self._lock:
            self._seq += 1
            return f"ord-{self._seq}"


def _fire(handler: QuoteHandler, q: Quote) -> None:
    """Invoke an async quote handler fire-and-forget, swallowing any exception —
    mirroring the C# ``try { _ = s(q); } catch { ... }``. If a running event loop
    is present the coroutine is scheduled on it; otherwise it is driven to
    completion on a throwaway loop so a synchronous caller still delivers.
    """
    try:
        result = handler(q)
    except Exception:  # noqa: BLE001 — a throwing subscriber must not break publish
        return
    if not asyncio.iscoroutine(result):
        return
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        loop = None
    if loop is not None:
        task = loop.create_task(result)
        # Swallow a later handler exception so it never surfaces as an unretrieved
        # task exception (matches the C# fire-and-forget try/catch).
        task.add_done_callback(_drain_task_exception)
    else:
        try:
            asyncio.run(result)
        except Exception:  # noqa: BLE001
            return


def _drain_task_exception(task: "asyncio.Task[None]") -> None:
    if task.cancelled():
        return
    # Retrieving the exception marks it as handled.
    task.exception()
